src/Adapter/MSTest.TestAdapter/Execution/TestAssemblyInfo.cs

```diff
@@ -251,36 +269,57 @@ internal void ExecuteAssemblyCleanup()
             return;
         }
 
+        Exception? assemblyCleanupException;
         lock (_assemblyInfoExecuteSyncObject)
         {
             try
             {
-                AssemblyCleanupMethod.InvokeAsSynchronousTask(null);
+                assemblyCleanupException = MethodRunner.RunWithTimeoutAndCancellation(
+                     () => AssemblyCleanupMethod.InvokeAsSynchronousTask(null),
+                     new CancellationTokenSource(),
+                     AssemblyCleanupMethodTimeoutMilliseconds,
+                     AssemblyCleanupMethod,
+                     Resource.AssemblyCleanupWasCancelled,
+                     Resource.AssemblyCleanupTimedOut);
             }
             catch (Exception ex)
             {
-                var realException = ex.GetRealException();
-
-                // special case AssertFailedException to trim off part of the stack trace
-                string errorMessage = realException is AssertFailedException or AssertInconclusiveException
-                    ? realException.Message
-                    : realException.GetFormattedExceptionMessage();
-
-                var exceptionStackTraceInfo = realException.GetStackTraceInformation();
-                DebugEx.Assert(AssemblyCleanupMethod.DeclaringType?.Name is not null, "AssemblyCleanupMethod.DeclaringType.Name is null");
-
-                throw new TestFailedException(
-                    UnitTestOutcome.Failed,
-                    string.Format(
-                        CultureInfo.CurrentCulture,
-                        Resource.UTA_AssemblyCleanupMethodWasUnsuccesful,
-                        AssemblyCleanupMethod.DeclaringType.Name,
-                        AssemblyCleanupMethod.Name,
-                        errorMessage,
-                        exceptionStackTraceInfo?.ErrorStackTrace),
-                    exceptionStackTraceInfo,
-                    realException);
+                assemblyCleanupException = ex;
             }
         }
+
+        // If assemblyCleanup was successful, then don't do anything
+        if (assemblyCleanupException is null)
+        {
+            return;
+        }
+
+        // Cache and return an already created TestFailedException.
+        if (assemblyCleanupException is TestFailedException)
```

Comment:
I am not sure to understand the logic behind this caching, could you please add a comment in the code.

Reply:
in the rest of the function we are creating a `TestFailedException` but in case the helper function returned this error I don't have to take time creating it again

Reply:
I think that the comment is confusing it's not a cache we simply throw it as is.

Reply:
updated the comment